### ISSUE TITLE
revisit httptest

### ISSUE THREAD

**Issue by gergness** [Crunch-io/rcrunch#450]
- learn from https://ropensci.org/technotes/2019/12/11/http-testing/ ?

- Make `gsub_request()` more targeted (had problem with crunchtabs because it trimmed a UUID from a POST)

**Comment by gergness** [comment]
- I'm still confused about redacting after working with crunchtabs. Seems like we should have a `httptest::gsub_request()` given that we redact responses? This has been a major pain.
- Also would be helpful if httptest kept a record of what the digest from the query string was from. For the POST files that get saved as .R files, it could be comments, for json files maybe it would require a parallel file? This would require work on httptest

**Comment by 1beb** [comment]
Remember to document gsub for post/body as work around for POST problems with http test redaction. 

Example: 

```
with_api_fixture <- function(fixture_path, expr) {
  with(
    crunch::temp.options(
      crunch.api = "https://app.crunch.io/api/",
      httptest.mock.paths = fixture_path,
      crunch.show.progress = FALSE
    ),
    httptest::with_mock_api(
      # Also need to redact UUIDs as is done in POSTs
      with_mock(
        `crunch::crPOST` = function(...) {
          args <- list(...)
          # Necessary for post
          args$body <- gsub("([0-9a-f]{6})[0-9a-f]{26}", "\\1", args$body)
          args[[1]] <- gsub("([0-9a-f]{6})[0-9a-f]{26}", "\\1", args[[1]])
          do.call(
            function(...) crunch:::crunchAPI("POST", ...),
            args
          )
        },
        expr
      )
    )
  )
}
```

**Comment by gergness** [comment]
> @1beb: Can we make with_api_fixutre an official part of rcrunch utils so I don't have to copy over?

https://github.com/Crunch-io/crunchtabs/blob/450b5df160538269b6c0cdf183b4c6cb27952ef7/R/utils.R#L165

**Comment by 1beb** [comment]
Also need to document dataset calls so people know dataset length should be truncated. Something like this: 

```
#' When building a fixture that relies on a dataset identified by a link that uses the dataset #' id. Such as https://app.crunch.io/datasets/:id: when calling that link within your 
#' with_api_fixture expression, you should truncate the id to the first 6 characters. So an id #' value like 1jk23hldfl1u23 becomes 1jk23h instead.
#'
```
Once we get this set, I think it's a good idea to create a vignette on how to create fixtures and usage examples of with_api_fixture. I can assist with that.

**Comment by 1beb** [comment]
@malecki becoming more pressing as a larger range of individuals are starting to use this. Can we include with_api_fixture code in the rcrunch repo? Right now I have it as a util in crunchtabs but I think that leaves room for unnecessary divergence in the future.

**Comment by 1beb** [comment]
Closed by #494
